Fixed #24018 -- Allowed setting pragma options on SQLite. (#14824)
django/django: docs/ref/databases.txt

.. _oracle-notes:

> Setting PRAGMA Options

Author: Apologies, I actually ran `make html` this time and double-checked the output.

.. _JSON1 extension: https://www.sqlite.org/json1.html
.. _the wiki page: https://code.djangoproject.com/wiki/JSON1Extension

.. _sqlite-init-command:

> Setting PRAGMA Options

Reviewer: ```suggestion
Setting PRAGMA options
```

Reviewer: As far as I'm aware, `pragma` should be uppercased only when we talk about an SQL statement.
```suggestion
Setting pragma options
```

Author: Corrected

----------------------

> .. versionadded:: 5.0

Reviewer 2: ```suggestion
.. versionadded:: 5.1
```

Author: Fixed, thanks!

> Pragma options can be set upon connection by using the ``init_command`` option

Reviewer: I'd add a link to the `https://www.sqlite.org/pragma.html`
```suggestion
:ref:`Pragma options <https://www.sqlite.org/pragma.html>` can be set upon connection by using the ``init_command`` option
```
Remember to re-wrap at 79 chars.

Author: I'm a bit new to Sphinx; when I did that exact line I was getting a `WARNING: undefined label` error. I added it where the link reference is defined separately (this seems to be similar to how it is done elsewhere in this file).

> similar to the MySQL backend. The example below shows how to disable

Reviewer: `similar to the MySQL backend.` is unnecessary. There is not much description in the MySQL docs that could be helpful for SQLite users.

Author: Removed.

> Pragma options can be set upon connection by using the ``init_command`` option
> similar to the MySQL backend. The example below shows how to disable
> synchronous writes::
> 
>     DATABASES = {
>         "default": {
>             "ENGINE": "django.db.backends.sqlite3",
>             "NAME": BASE_DIR / "db.sqlite3",
>             "OPTIONS": {
>                 "init_command": "PRAGMA synchronous=0;",

Reviewer 3: Can we change the example to `journal_mode=WAL` instead of `synchronous=0`? `synchronous=0` could corrupt the database on system crash/power loss, so mindlessly copy-pasting the example could lead to data loss.

I'd prefer to see `journal_mode=WAL` here since I think it will be a common thing to set, but `synchronous=3` would also be okay since it increases durability.

```suggestion
Pragma options can be set upon connection by using the ``init_command`` option
similar to the MySQL backend. The example below shows how to enable WAL
journaling mode::

    DATABASES = {
        "default": {
            "ENGINE": "django.db.backends.sqlite3",
            "NAME": BASE_DIR / "db.sqlite3",
            "OPTIONS": {
                "init_command": "PRAGMA journal_mode=WAL;",
```

Author: There was some [feedback in trac](https://code.djangoproject.com/ticket/24018#comment:8) that noted that since `journal_mode=WAL` is persistent, that setting it once via the CLI would be sufficient.

I don't have a strong preference (the original PR used WAL in the example) and will defer to others on this one (and agree on reasons for showing sync=3 instead of sync=0 if that option is preferred too).

Reviewer: Yes, please change to the `synchronous=3`.

Author: Done
